kong/cmd/utils/process.lua

```diff
@@ -0,0 +1,316 @@
+    -- the input as a numeric string
+    pid, err = pid_from_file(target)
+
+    -- PID was supplied as a numeric string (i.e. "123")
+    if err == E_NO_SUCH_PID_FILE and validate_pid(target) then
+      pid = validate_pid(target)
+      err = nil
+    end
+
+  else
+    error(E_INVALID_PID_TYPE .. ": '" .. typ .. "'", 2)
+  end
+
+  if not pid then
+    return nil, err
+
+  elseif not validate_pid(pid) then
+    error(E_INVALID_PID, 2)
+  end
+
+  return pid
+end
+
+
+---
+-- Target processes may be referenced by their integer id (PID)
+-- or by a pid filename.
+--
+---@alias kong.cmd.utils.process.target
+---| integer # pid
+---| string  # pid file
+
+
+---
+-- Send a signal to a process.
+--
+-- The signal may be specified as a name ("TERM") or integer (15).
+--
+---@param  target      kong.cmd.utils.process.target
+---@param  sig         resty.signal.name|integer
+---@return boolean|nil ok
+---@return nil|string  error
+local function signal(target, sig)
+  local pid, err = get_pid(target)
+
+  if not pid then
+    return nil, err
+  end
+
+  return resty_kill(pid, sig)
+end
+
+
+---
+-- Send the TERM signal to a process.
+--
+---@param  target      kong.cmd.utils.process.target
+---@return boolean|nil ok
+---@return nil|string  error
+local function term(target)
+  return signal(target, SIG_TERM)
+end
+
+
+---
+-- Send the KILL signal to a process.
+--
+---@param  target      kong.cmd.utils.process.target
+---@return boolean|nil ok
+---@return nil|string  error
+local function kill(target)
+  return signal(target, SIG_KILL)
+end
+
+
+---
+-- Send the QUIT signal to a process.
+--
+---@param  target      kong.cmd.utils.process.target
+---@return boolean|nil ok
+---@return nil|string  error
+local function quit(target)
+  return signal(target, SIG_QUIT)
+end
+
+
+---
+-- Send the HUP signal to a process.
+--
+---@param  target      kong.cmd.utils.process.target
+---@return boolean|nil ok
+---@return nil|string  error
+local function hup(target)
+  return signal(target, SIG_HUP)
+end
+
+
+---
+-- Check for the existence of a process.
+--
+-- Under the hood this sends the special `0` signal to check the process state.
+--
+-- Returns:
+--   * true|false under normal circumstances
+--   * nil+error for invalid input
+--
+-- Throws for unexpected errors from resty.signal.
+--
+-- Callers should decide for themselves how strict they must be when handling
+-- errors. For instance, when NGINX is starting up there is a period where the
+-- pidfile may be empty or non-existent, which will result in this function
+-- returning nil+error. For some callers this might be expected and acceptible,
+-- but for others it may not.
+--
+---@param  target      kong.cmd.utils.process.target
+---@return boolean|nil exists
+---@return nil|string  error
+local function exists(target)
+  local pid, err = get_pid(target)
+  if not pid then
+    return nil, err
+  end
+
+  local ok
+  ok, err = resty_kill(pid, 0)
+
+  if ok then
```

Comment:
I would find it surprising for an `exists` function to error if the process exists, but cannot be managed.  I would expect it to return `true` instead.

Reply:
We talked about this outside of GH, but I'm continuing the discussion here for the sake of anyone else who wants to chime in. I'd like some consensus before merging.

---

In a vacuum, returning `true` seems to be the most correct. In our case, this _does_ have user-facing side effects.

The primary example is in the `kong-health` script. Start kong as root and then run `kong-health` as an unprivileged user.

In `master`, `kong-health` reports that kong is not running because of the current behavior of the `is_running()` API:

```
$ kong-health

Error: Kong is not running at /home/michaelm/git/kong/kong/eperm

  Run with --v (verbose) or --vv (debug) for more details
```

Whereas changing `is_running()` (now `exists()`) to return true in this condition means that this branch will report that kong is running:

```
$ kong-health

Kong is healthy at /home/michaelm/git/kong/kong/eperm
```

I would say this is arguably more "correct", but it may be a surprising change to introduce.